ID review suggestions (#12)
OpenLiberty/guide-grpc-intro: README.adoc

Protocol buffers allow cross-project support through the `.proto` file. As a result, gRPC clients and servers can run and communicate with each other from different environments. For example, a gRPC client running in Java codebase is able to call a gRPC server from other https://grpc.io/docs/languages/[supported languages^]. This feature of protocol buffers allows for easier integration between services.

== What you'll learn

You will learn how to create gRPC services and their clients by using protocol buffers and how to implement them with Open Liberty. You will use Maven to generate the gRPC stubs, deploy the services, and to interact with the running Liberty server.

The application that you will build in this guide consists of the `systemproto` model project, the `query` client service, and the `system` server service. The `query` service implements four RESTful APIs by using four different gRPC streaming methods.

> 

Author: added a detail of the four RPCs to orient the reader, please review for accuracy

* Unary RPC: The client sends a single request and recieves a single response.

> * Server streaming RPC: The client sends and single request and the server returns a stream of messages.

Author: ```suggestion
* Server streaming RPC: The client sends a single request and the server returns a stream of messages.
```

* Client streaming RPC: The client sends a stream of messages and the server responds with a single message.

> * Bidirectional RPC: Both client and server sned a stream of messages and the client and server can read and write messages in any order.

Author: ```suggestion
* Bidirectional RPC: Both client and server send a stream of messages. The client and server can read and write messages in any order.
```

mvn -pl system liberty:run
----

> Next, open another command-line session, navigate to the `finish` directory, and start the `query` service by using the following command:

Author: Throughout, I suggest reminding the user to navigate to the working directory when you open a new terminal

// static guide instruction
ifndef::cloud-hosted[]

> Wait until you the `The defaultServer server is ready to run a smarter planet` message in your consoles. Then, point your browser to the http://localhost:9081/query/properties/os.name URL to test out the basic unary service. You will see your operating system name. Next, point your browser to the following URLs to try out the corresponding streaming rpc call: 

Author: ```suggestion
Wait until you see the `The defaultServer server is ready to run a smarter planet` message in your consoles. Then, point your browser to the http://localhost:9081/query/properties/os.name URL to test out the basic unary service. You will see your operating system name. Next, point your browser to the following URLs to try out the corresponding streaming rpc call: 
```

ifndef::cloud-hosted[]

> Point your browser to the http://localhost:9081/query/properties/os.name URL to test out the basic unary service. You will see your operating system name. Next, point your browser to the following URLs to try out the corresponding streaming rpc call: 

Author: ```suggestion
Wait until you the `The defaultServer server is ready to run a smarter planet` message in your consoles. Then, point your browser to the http://localhost:9081/query/properties/os.name URL to test out the basic unary service. You will see your operating system name. Next, point your browser to the following URLs to try out the corresponding streaming rpc call: 
```

include::finish/systemproto/pom.xml[]
----

The first few lines define the [hotspot=basicConfig file=0]`syntax`, [hotspot=basicConfig file=0]`package`, and [hotspot=basicConfig file=0]`option` basic configuration of the `.proto` file. The [hotspot=SystemService file=0]`SystemService` service contains the four service calls that you will implement in the coming sections.

The [hotspot=getProperty file=0]`getProperty` rpc defines the unary call. In this call, the client side sends a `SystemPropertyName` message to the server side, which returns a `SystemPropertyValue` message with the property value. The [hotspot=SystemPropertyName file=0]`SystemPropertyName` and [hotspot=SystemPropertyValue file=0]`SystemPropertyValue` message types define that the `propertyName` and `propertyValue` must be string.

> The [hotspot=getPropertiesServer file=0]`getPropertiesServer` rpc defines the server streaming call. The client side sends a `SystemPropertyPrefix` message to the server side. The server returns a stream of `SystemProperty` messages. Each [hotspot=SystemProperty file=0]`SystemProperty` message contains `propertyName` and `propertyValue` strings.

Author: ```suggestion
The [hotspot=getPropertiesServer file=0]`getPropertiesServer` RPC defines the server streaming call. The client side sends a `SystemPropertyPrefix` message to the server side. The server returns a stream of `SystemProperty` messages. Each [hotspot=SystemProperty file=0]`SystemProperty` message contains `propertyName` and `propertyValue` strings.
```

> The [hotspot=getPropertiesClient file=0]`getPropertiesClient` rpc defines the client streaming call. The client side streams `SystemPropertyName` messages to the server side. The server returns a [hotspot=SystemProperties file=0]`SystemProperties` message that contains a map of the properties with their respective values.

Author: ```suggestion
The [hotspot=getPropertiesClient file=0]`getPropertiesClient` RPC defines the client streaming call. The client side streams `SystemPropertyName` messages to the server side. The server returns a [hotspot=SystemProperties file=0]`SystemProperties` message that contains a map of the properties with their respective values.
```

> The [hotspot=getPropertiesBidirect file=0]`getPropertiesBidirect` rpc defines the bidirectional streaming call. In this service, the client side streams `SystemPropertyName` messages to the server side. The server returns a stream of [hotspot=SystemProperty file=0]`SystemProperty` messages.

Author: ```suggestion
The [hotspot=getPropertiesBidirect file=0]`getPropertiesBidirect` RPC defines the bidirectional streaming call. In this service, the client side streams `SystemPropertyName` messages to the server side. The server returns a stream of [hotspot=SystemProperty file=0]`SystemProperty` messages.
```

include::finish/system/src/main/java/io/openliberty/guides/system/SystemService.java[]
----

The `SystemService` class implements the `SystemServiceGrpc` class that is generated by the `.proto` file. The four types of services defined in the proto file are implemented in this class.

> The [hotspot=getProperty file=0]`getProperty()` method implements the rpc unary call defined in the `.proto` file. The [hotspot=pName file=0]`getPropertyName()` getter method that is generated by gRPC retrieves the property name from the client, and stores it into the `pName` variable. The System property value is stored into the [hotspot=pValue file=0]`pValue` variable. The gRPC APIs wil create a [hotspot=response file=0]`SystemPropertyValue` message, with its type defined in the `SystemService.proto` file. Then, the message is sent to the client service through the `StreamObserver` by using its [hotspot=next file=0]`onNext()` and [hotspot=complete file=0]`onComplete()` methods.

Author: ```suggestion
The [hotspot=getProperty file=0]`getProperty()` method implements the rpc unary call defined in the `.proto` file. The [hotspot=pName file=0]`getPropertyName()` getter method that is generated by gRPC retrieves the property name from the client, and stores it into the `pName` variable. The System property value is stored into the [hotspot=pValue file=0]`pValue` variable. The gRPC APIs will create a [hotspot=response file=0]`SystemPropertyValue` message, with its type defined in the `SystemService.proto` file. Then, the message is sent to the client service through the `StreamObserver` by using its [hotspot=next file=0]`onNext()` and [hotspot=complete file=0]`onComplete()` methods.
```

include::finish/system/src/main/liberty/config/server.xml[]
----

> Add the [hotspot=grpc file=1]`grpc` feature to the Liberty server configuration file. This feature enables compatibility between gRPC server and the Liberty server. Configure the [hotspot=grpcConfig file=1]`grpc` feature with the `target` attribute. If you want to learn more about `grpc` feature configuration, see the https://openliberty.io/docs/latest/reference/config/grpc.html[GRPC Server Properties^].

Author: ```suggestion
Add the [hotspot=grpc file=1]`grpc` feature to the Liberty server configuration file. This feature enables compatibility between the gRPC server and the Liberty server. Configure the [hotspot=grpcConfig file=1]`grpc` feature with the `target` attribute. If you want to learn more about `grpc` feature configuration, see the https://openliberty.io/docs/latest/reference/config/grpc.html[GRPC Server Properties^].
```

----
include::finish/query/src/main/java/io/openliberty/guides/query/PropertiesResource.java[]
----

> The `PropertiesResource` class provides RESTful endpoints to interact with the `system` service. The [hotspot=unary file=2]`/query/properties/${property}` endpoint uses the unary service call to get the property value from the `system` service. The endpoint creates a [hotspot=createChannel1 file=2]`channel`, which it uses to create a client by the [hotspot=createClient1 file=2]`SystemServiceGrpc.newBlockingStub()` API. The endpoiint then uses the client to get the property value, shuts down the channel, and immediately returns the value from the `system` service response.

Author: ```suggestion
The `PropertiesResource` class provides RESTful endpoints to interact with the `system` service. The [hotspot=unary file=2]`/query/properties/${property}` endpoint uses the unary service call to get the property value from the `system` service. The endpoint creates a [hotspot=createChannel1 file=2]`channel`, which it uses to create a client by the [hotspot=createClient1 file=2]`SystemServiceGrpc.newBlockingStub()` API. The endpoint then uses the client to get the property value, shuts down the channel, and immediately returns the value from the `system` service response.
```

> The `PropertiesResource` class provides RESTful endpoints to interact with the `system` service. The [hotspot=unary file=2]`/query/properties/${property}` endpoint uses the unary service call to get the property value from the `system` service. The endpoint creates a [hotspot=createChannel1 file=2]`channel`, , which it uses to create a client by the [hotspot=createClient1 file=2]`SystemServiceGrpc.newBlockingStub()` API. The endpoiint then uses the client to get the property value, shuts down the channel, and immediately returns the value from the `system` service response.

Author: ```suggestion
The `PropertiesResource` class provides RESTful endpoints to interact with the `system` service. The [hotspot=unary file=2]`/query/properties/${property}` endpoint uses the unary service call to get the property value from the `system` service. The endpoint creates a [hotspot=createChannel1 file=2]`channel`, which it uses to create a client by the [hotspot=createClient1 file=2]`SystemServiceGrpc.newBlockingStub()` API. The endpoiint then uses the client to get the property value, shuts down the channel, and immediately returns the value from the `system` service response.
```

> The `PropertiesResource` class provides RESTful endpoints to interact with the `system` service. The [hotspot=unary file=2]`/query/properties/${property}` endpoint uses the unary service call to get the property value from the `system` service. The endpoint creates a [hotspot=createChannel1 file=2]`channel`, , which it uses to create a client by the [hotspot=createClient1 file=2]`SystemServiceGrpc.newBlockingStub()` API. The endpoiint then uses the client to get the property value, shutdowns the channel, and then immediately returns the value from the `system` service response.

Author: ```suggestion
The `PropertiesResource` class provides RESTful endpoints to interact with the `system` service. The [hotspot=unary file=2]`/query/properties/${property}` endpoint uses the unary service call to get the property value from the `system` service. The endpoint creates a [hotspot=createChannel1 file=2]`channel`, , which it uses to create a client by the [hotspot=createClient1 file=2]`SystemServiceGrpc.newBlockingStub()` API. The endpoiint then uses the client to get the property value, shuts down the channel, and immediately returns the value from the `system` service response.
```

include::finish/query/src/main/liberty/config/server.xml[]
----

Add the [hotspot=grpc file=3]`grpc` and [hotspot=grpcClient file=3]`grpcClient` features to the Liberty query server configuration file. Both features enable gRPC server and client compatibility with the Liberty server. Configure the [hotspot=grpcClientConfig file=3]`grpcClient` feature with the `headersToPropagate` and `target` attributes. If you want to learn more about `grpcClient` feature configuration, see the https://openliberty.io/docs/latest/reference/config/grpcClient.html[GRPC Client Properties^].

Because you are running the `system` and `query` services in dev mode, the changes that you made were automatically picked up. You’re now ready to check out your application in your browser.

// static guide instruction
ifndef::cloud-hosted[]
Point your browser to the http://localhost:9081/query/properties/os.name URL to test out the unary service. Your operating system name is displayed. You can try out the URL with other property name.
endif::[]
// cloud hosted instruction
ifdef::cloud-hosted[]

> Click the following button to visit the ***/query/properties/os.name*** endpoint to test out the unary service. Your operating system name is displayed. You can also try out the URL with other property names.

Author: I tried the other property names and got an exception- which I believe is expected because this is a unary call= 1 property only.

If that's the case we should clarify

Reviewer: See comment below for resolution

> Click the following button to visit the ***/query/properties/os.name*** endpoint to test out the unary service. Your operating system name is displayed. You can also try out the URL with other property names.

Author: ```suggestion
Click the following button to visit the ***/query/properties/os.name*** endpoint to test out the unary service. Your operating system name is displayed. You can also try out the URL with other property names but they will return an exception because the unary RPC only returns a single property.
```

Reviewer: After discussion, we think it is better to leave the last sentence out to clear up confusion ("properties" is being used in two different contexts within the guide- it means os properties but also system properties).

Addressed in https://github.com/OpenLiberty/draft-guide-grpc-intro/pull/12/commits/69cf4819db94af6e8c3e92b65e5084f2be711e15.

> Click the following button to visit the ***/query/properties/os.name*** endpoint to test out the unary service. Your operating system name is displayed.. You can also try out the URL with other property names.

Author: ```suggestion
Click the following button to visit the ***/query/properties/os.name*** endpoint to test out the unary service. Your operating system name is displayed. You can also try out the URL with other property names.
```

// Implementing server streaming call
// ===============================================================================================

== Implementing the server streaming call

> In the server streaming call, the `query` client service provides the `/query/properties/os` endpoint that sends a message to the `system` server service. The `system` service streams the properties started with `os.` back to the `query` service. A channel is created between the `query` and the `system` services to stream messages. The channel is closed by the `system` service only after sending the last message to the `query` service. 

Author: ```suggestion
In the server streaming call, the `query` client service provides the `/query/properties/os` endpoint that sends a message to the `system` server service. The `system` service streams any properties that start with `os.` back to the `query` service. A channel is created between the `query` and the `system` services to stream messages. The channel is closed by the `system` service only after sending the last message to the `query` service. 
```

include::finish/system/src/main/java/io/openliberty/guides/system/SystemService.java[]
----

> The [hotspot=getPropertiesServer file=0]`getPropertiesServer()` method implements server streaming rpc call. The [hotspot=prefix file=0]`getPropertyPrefix()` getter method retrieves the property prefix from the client. Properties that start with the [hotspot=filter file=0]`prefix` are filtered out. For each property, a [hotspot=serverMessage file=0]`SystemProperty` message is built and streamed to the client through the `StreamObserver` by using its [hotspot=serverNext1 file=0]`onNext()` method. When all properties are streamed, the service stops streaming by calling the [hotspot=serverComplete file=0]`onComplete()` method.

Author: ```suggestion
The [hotspot=getPropertiesServer file=0]`getPropertiesServer()` method implements the server streaming RPC call. The [hotspot=prefix file=0]`getPropertyPrefix()` getter method retrieves the property prefix from the client. Properties that start with the [hotspot=filter file=0]`prefix` are filtered out. For each property, a [hotspot=serverMessage file=0]`SystemProperty` message is built and streamed to the client through the `StreamObserver` by using its [hotspot=serverNext1 file=0]`onNext()` method. When all properties are streamed, the service stops streaming by calling the [hotspot=serverComplete file=0]`onComplete()` method.
```

include::finish/query/src/main/java/io/openliberty/guides/query/PropertiesResource.java[]
----

> The endpoint creates a [hotspot=createChannel2 file=1]`channel` to the `system` service and a [hotspot=createClient2 file=1]`client` by using the `SystemServiceGrpc.newStub()` API. Then, it calls the [hotspot=getPropertiesServer file=1]`getPropertiesServer()` method with an implementation of the `StreamObserver` interface. The [hotspot=onNext1 file=1]`onNext()` method receives messages streaming from the server individually and stores them into the `properties` placeholder. After all properties are received, it shuts down the [hotspot=closeConnection file=1]`channel` and returns the placeholder. Because the rpc call is asynchronous, a [hotspot=countDownLatch1 hotspot=countDownLatch2 hotspot=countDownLatch3 file=1]`CountDownLatch` synchronizes the streaming flow.

Author: ```suggestion
The endpoint creates a [hotspot=createChannel2 file=1]`channel` to the `system` service and a [hotspot=createClient2 file=1]`client` by using the `SystemServiceGrpc.newStub()` API. Then, it calls the [hotspot=getPropertiesServer file=1]`getPropertiesServer()` method with an implementation of the `StreamObserver` interface. The [hotspot=onNext1 file=1]`onNext()` method receives messages streaming from the server individually and stores them into the `properties` placeholder. After all properties are received, it shuts down the [hotspot=closeConnection file=1]`channel` and returns the placeholder. Because the rpc call is asynchronous, a [hotspot=countDownLatch1 hotspot=countDownLatch2 hotspot=countDownLatch3 file=1]`CountDownLatch` instance synchronizes the streaming flow.
```

> The endpoint creates a [hotspot=createChannel2 file=1]`channel` to the `system` service and a [hotspot=createClient2 file=1]`client` by using the `SystemServiceGrpc.newStub()` API. Then, it calls the [hotspot=getPropertiesServer file=1]`getPropertiesServer()` method with an implementation of the `StreamObserver` interface. The [hotspot=onNext1 file=1]`onNext()` method receives messages streaming from the server individually and stores them into the `properties` placeholder. After all properties are received, it shuts down the [hotspot=closeConnection file=1]`channel` and returns the placeholder. Because the rpc call is asynchronous, a [hotspot=countDownLatch1 hotspot=countDownLatch2 hotspot=countDownLatch3 file=1]`CountDownLatch` instance synchronizes the streaming flow.

Author: At the beginning of this section, it says 

> "The channel is only closed _by the system service_ after sending the last message to the query service.

But here it sounds like the query service shuts down the channel...

> After all properties are received, it shutdowns the channel...

Should it be 

> After all properties are received, the system service shuts down the channel...

Reviewer: Absolutely .... 
Committed the changes at https://github.com/OpenLiberty/draft-guide-grpc-intro/pull/12/commits/083a029ca2c938993e0ae3649be2e2dfee5067e8

> The endpoint creates a [hotspot=createChannel2 file=1]`channel` to the `system` service and a [hotspot=createClient2 file=1]`client` by using the `SystemServiceGrpc.newStub()` API. Then, it calls the [hotspot=getPropertiesServer file=1]`getPropertiesServer()` method with an implementation of the `StreamObserver` interface. The [hotspot=onNext1 file=1]`onNext()` method receives messages streaming from the server individually and stores them into the `properties` placeholder. After all properties are received, the `system` service shuts down the [hotspot=closeConnection file=1]`channel` and returns the placeholder. Because the rpc call is asynchronous, a [hotspot=countDownLatch1 hotspot=countDownLatch2 hotspot=countDownLatch3 file=1]`CountDownLatch` instance synchronizes the streaming flow.

Author: ```suggestion
The endpoint creates a [hotspot=createChannel2 file=1]`channel` to the `system` service and a [hotspot=createClient2 file=1]`client` by using the `SystemServiceGrpc.newStub()` API. Then, it calls the [hotspot=getPropertiesServer file=1]`getPropertiesServer()` method with an implementation of the `StreamObserver` interface. The [hotspot=onNext1 file=1]`onNext()` method receives messages streaming from the server individually and stores them into the `properties` placeholder. After all properties are received, the `system` service shuts down the [hotspot=closeConnection file=1]`channel` and returns the placeholder. Because the RPC call is asynchronous, a [hotspot=countDownLatch1 hotspot=countDownLatch2 hotspot=countDownLatch3 file=1]`CountDownLatch` instance synchronizes the streaming flow.
```

// ===============================================================================================

== Implementing the client streaming call

In the client streaming call, the `query` client service provides the `/query/properties/user` endpoint, which streams the user properties to the `system` server service. The `system` service returns a map of user properties with their values.

> Update the `SystemService` class to implement the client streaming rpc call.

Author: ```suggestion
Update the `SystemService` class to implement the client streaming RPC call.
```

> == Implementing bidirectional streaming call

Author: ```suggestion
== Implementing the bidirectional streaming call
```

----
include::finish/query/src/test/java/it/io/openliberty/guides/query/QueryIT.java[]
----

> Each test case tests one of the methods for instantiating a gRPC client.

Author: are they RESTful clients or gRPC clients?

Reviewer: It should be gRPC clients, I believe ....

== Great work! You're done!

> You just developed an application that implements four gRPC streaming calls with Open Liberty. For more information about using gRPC with Open Liberty, see https://openliberty.io/docs/latest/grpc-services.html[Provide and consume gRPC services on Open Liberty] in the Open Liberty docs.

Author: ```suggestion
You just developed an application that implements four gRPC streaming calls with Open Liberty. For more information, see https://openliberty.io/docs/latest/grpc-services.html[Provide and consume gRPC services on Open Liberty] in the Open Liberty docs.
```
